fuel-vm/src/tests/coins.rs

```diff
@@ -0,0 +1,884 @@
+        op::log(RegId::ZERO, RegId::ZERO, RegId::ZERO, RegId::ZERO),
+        op::ret(RegId::ONE),
+    ]);
+
+    let mut test_context = TestBuilder::new(1234u64);
+    let contract_id = test_context
+        .setup_contract(ops, Some((AssetId::zeroed(), Word::MAX - helper)), None) // Ensure enough but not too much balance
+        .contract_id;
+    assert_eq!(extract_novalue(&run(test_context, contract_id)), result);
+}
+
+type MintOrBurnOpcode = fn(RegId, RegId) -> Instruction;
+
+#[test_case(vec![(op::mint, 0, 0)] => RunResult::Success(()); "Mint 0")]
+#[test_case(vec![(op::burn, 0, 0)] => RunResult::Success(()); "Burn 0")]
+#[test_case(vec![(op::mint, 100, 0)] => RunResult::Success(()); "Mint 100")]
+#[test_case(vec![(op::mint, Word::MAX, 0)] => RunResult::Success(()); "Mint Word::MAX")]
+#[test_case(vec![(op::mint, 100, 0), (op::burn, 100, 0)] => RunResult::Success(()); "Mint 100, Burn all")]
+#[test_case(vec![(op::mint, Word::MAX, 0), (op::burn, Word::MAX, 0)] => RunResult::Success(()); "Mint Word::MAX, Burn all")]
+#[test_case(vec![(op::mint, 100, 0), (op::mint, 10, 0), (op::burn, 20, 0)] => RunResult::Success(()); "Mint 10 and 10, Burn all")]
+#[test_case(vec![(op::mint, 2, 0), (op::mint, 3, 1), (op::burn, 2, 0), (op::burn, 3, 1)] => RunResult::Success(()); "Mint multiple assets, Burn all")]
+#[test_case(vec![(op::burn, 1, 0)] => RunResult::Panic(PanicReason::NotEnoughBalance); "Burn nonexisting 1")]
+#[test_case(vec![(op::burn, Word::MAX, 0)] => RunResult::Panic(PanicReason::NotEnoughBalance); "Burn nonexisting Word::MAX")]
+#[test_case(vec![(op::mint, Word::MAX, 0), (op::mint, 1, 0)] => RunResult::Panic(PanicReason::BalanceOverflow); "Mint overflow")]
+#[test_case(vec![(op::mint, Word::MAX, 0), (op::burn, 1, 0), (op::mint, 2, 0)] => RunResult::Panic(PanicReason::BalanceOverflow); "Mint,Burn,Mint overflow")]
+fn mint_burn_single_sequence(seq: Vec<(MintOrBurnOpcode, Word, u8)>) -> RunResult<()> {
+    let reg_len: u8 = 0x10;
+    let reg_mint_amount: u8 = 0x11;
+
+    let mut ops = vec![
+        // Allocate space for sub asset id
+        op::movi(reg_len, 32),
+        op::aloc(reg_len),
+    ];
+
+    for (mint_or_burn, amount, sub_id) in seq {
+        ops.push(op::sb(RegId::HP, sub_id, 0));
+        ops.extend(set_full_word(reg_mint_amount.into(), amount));
+        ops.push(mint_or_burn(reg_mint_amount.into(), RegId::HP));
+    }
+    ops.push(op::ret(RegId::ONE));
+
+    let mut test_context = TestBuilder::new(1234u64);
+    let contract_id = test_context.setup_contract(ops, None, None).contract_id;
+    extract_novalue(&run(test_context, contract_id))
+}
+
+#[test_case(vec![false] => RunResult::Panic(PanicReason::NotEnoughBalance); "Burn")]
+#[test_case(vec![true, false, false] => RunResult::Panic(PanicReason::NotEnoughBalance); "Mint,Burn,Burn")]
+#[test_case(vec![true, true, false] => RunResult::Success(1); "Mint,Mint,Burn")]
+fn mint_burn_many_calls_sequence(seq: Vec<bool>) -> RunResult<Word> {
```

Comment:
How is this functionally different from the previous tests: `mint_burn_single_sequence`? I feel like it's just a variation on the same tests.

If we do keep it, I'd prefer to have some type 
```
enum MintOrBurn {
    Mint,
    Burn,
}
```
instead of `bool`. The `bool` is cardinally identical to `MintOrBurn`, but semantically confusing.

Reply:
This does them in seperate calls, instead of having single contract to do all the calls, testing that the results persist over multiple calls correctly.

I had exactly that type before, but the verbosity made the lines longer and thus harder to read. I can re-introduce that if you think it's better.

Reply:
I would prefer it, yeah. I prefer clarity to brevity.

Reply:
Fixed in 35050850